[DOC] Add traces panel procedures (#70204)
grafana/grafana: docs/sources/panels-visualizations/visualizations/traces/index.md

# Traces panel

Traces are data that lets you follow a request as it traverses the services in your infrastructure.

> The Traces panel visualizes this data into a diagram that allows you to easily interpret this data.

Author: Thank you for the suggestion. I like your wording suggestion, but I want to verify that the wording change doesn't change the technical meaning. 

I don't think passes through would be technically accurate because there may be places where the trace stops or encounters an error, so it technically wouldn't be passing through.

For more information about traces and how to use them, refer to the following documentation:

- [What are traces](/docs/grafana-cloud/traces)
- [Tracing in Explore]({{< relref "../../../explore/trace-integration/" >}})
- [Tempo data source]({{< relref "../../../datasources/tempo/" >}})

> - [Getting started with Tempo](/docs/tempo/latest/getting-started)

Author: The links on lines 21 and 24 are cross-repository links, so they don't use the relref formatting. I'll check them and see if I get a 404 on the local build.

> - [Getting started with Tempo](/docs/tempo/latest/getting-started)

Author: I updated the links. On a local build, the Getting tarted with Tempo doesn't work, but when I put the link into a real URL it works. I think we're good on these.

{{< figure src="/static/img/docs/explore/trace-view-9-4.png" class="docs-image--no-shadow" max-width= "900px" caption="Screenshot of the trace view" >}}

## Add a panel with tracing visualizations

> Once you have tracing data available in your Grafana instance, you can add tracing panels to your Grafana dashboards.

Reviewer: The stack represents the entire suite of Grafana Cloud products. We should probably call is 'hosted Grafana' instead.

> Once you have tracing data available in your Grafana instance, you can add tracing panels to your Grafana dashboards.

Author: @hedss  Thank you. I was going to ask for clarification from you. :)

Great catch, @imatwawana.

Using a dashboard variable, `traceID`, lets you create a query to show specific traces for a given trace ID.
For more information about dashboard variables, see the [Variables documentation]({{< relref "../../../dashboards/variables" >}}).

### Before you begin

To use this procedure, you need:

- A Grafana instance

> - A Tempo data source connected to your Grafana instance (see [documentation](/docs/grafana/datasources/tempo/))

Author: FIxed. Good catch.

### Add the Tracing panel query

> You need to add the tracing panel to your dashboard and define a query using the Panel editor.

Author: This is an introduction to what the procedure does. It's a synopsis of the general step specific for traces data.

> You need to add the tracing panel to your dashboard and define a query using the Panel editor.

Reviewer 2: Maybe if you added a desired outcome to the sentence it wouldn't feel like I've landed in the middle of paragraph. Perhaps something like:
```suggestion
To view and analyze traces data in a dashboard, you need to add the tracing panel to your dashboard and define a query using the panel editor.
```

> You need to add the tracing panel to your dashboard and define a query using the Panel editor.

Author: @imatwawana You're a genious! That's exactly what that sentence needed. Thank you so much for that. I kept reading that sentence and wasn't happy with it, so your suggestion helps a bunch.

The query determines the data that is displayed in the panel.
For more information on the Panel editor, see the [Panel editor]({{< relref "../../panel-editor-overview" >}}) documentation.

This procedure uses dashboard variables and templates to allow us to enter trace IDs which can then be visualized. We’re going to use a variable called `traceId` and add it as a template query.

1. From your Grafana instance, create a new dashboard or go to an existing dashboard where you would like to add tracing panels.
1. Select **Add visualization** from a new dashboard or select **Add Panel** on an existing dashboard.
1. Search for and select the appropriate tracing data source.
1. In the panel search pane on the right, select the **Visualizations** tab, search for and select for Traces.

> 1. On the **Panel editor**, enter a **Title** for your trace panel. For more information on the panel editor, see the [Panel editor documentation]({{< relref "../../panel-editor-overview" >}}).

Author: oOo! Excellent suggestion. I'll update the link in your suggestion to point to the panel options instead.

1. Select the **TraceQL** query type tab from the query editor.

> 1. Enter `${traceId}` in the TraceQL query field to create a dashboard variable. This variable is used as the template query. <br /> {{< figure src="/static/img/docs/panels/traces/screenshot-traces-template-query.png" caption="Add a template query" >}}

Reviewer 3: This looks like there is another way to fix this besides inserting HTML. What  is the indentation were two spaces?

Author: I have been trying to figure out another way, and so far using br has been the only thing that worked. I tried using two indents at first. Using indentation with two spaces caused the images to have a number.

Author: I wonder it this has something to do with using a shortcode instead of the markdown image syntax.

Author: Got this sorted with help from Jack!

1. Select **Apply** in the Panel editor to add the panel to the dashboard.

> 1. Expand the size of the trace panel so that it fills the width of the dashboard. We’ll now use the dashboard settings to create the variable used in the panel’s template query.

Author: Thank you for the explanation and sharing what your vision is for these panels and visualizations pages.

> 1. Go to the dashboard **Settings** and add a new variable called `traceId`, of variable type **Custom**, giving it a label if required. Select **Apply** to add the variable to the dashboard. <br /> {{< figure  src="/static/img/docs/panels/traces/screenshot-traces-custom-variable.png" maxWidth="25%" caption="Add a Custom variable" >}}

Author: Agreed. I fought with this image last night. The reason I used the figure shortcode was because it offers a maxWidth capability. Or at least it's supposed to. Unfortunately, no matter what I did, I couldn't control the size of the image. I could change to HTML mark up instead. If all else fails, I can resize the image and upload it again.

> 1. Go to the dashboard **Settings** and add a new variable called `traceId`, of variable type **Custom**, giving it a label if required. Select **Apply** to add the variable to the dashboard. <br /> {{< figure  src="/static/img/docs/panels/traces/screenshot-traces-custom-variable.png" maxWidth="25%" caption="Add a Custom variable" >}}

Author: Fixed. I should have used max-width instead of maxWidth. I will update the writers-toolkit.

1. In the same dashboard that we added the trace visualization, select **Add panel** to add a new visualization panel.
1. Select the same trace data source you used in the previous section.
1. In the panel search pane, select the Visualizations tab and search for **Table**.
1. Select the **TraceQL** tab from the query editor.
1. On the **Panel editor**, enter a **Title** for your trace panel.
1. Add an appropriate TraceQL query to search for traces that you would like to visualize in the dashboard. We’ll use a simple, static query as an example. You can write the TraceQL query as a template query to take advantage of other dashboard variables, if they exist. This lets you create dynamic queries based on these variables. <br /> {{< figure src="/static/img/docs/panels/traces/screenshot-traces-dynamic-query.png" caption="Create a dynamic query" >}}

When results are returned from a query, the results are rendered in the panel’s table.

{{< figure src="/static/img/docs/panels/traces/screenshot-traces-returned-query.png" caption="Results of a returned query in the panel table" >}}

### Use a variable to add other links to traces

> Traces results include links to an Explorer page that render the trace. You can add other links to traces in the table that fill in the `traceId` dashboard variable when selected so that the trace is visualized in the same dashboard.

Reviewer 2: ```suggestion
The results in the Traces panel include links to the **Explore** page that renders the trace. You can add other links to traces in the table that fill in the `traceId` dashboard variable when selected, so that the trace is visualized in the same dashboard.
```

> To do so, we’re going to create a set of data links in the panel.

Reviewer 2: ```suggestion
To do so, you need to create a set of data links in the panel. Take the following steps:
```

> 1. Select **Add link** from the **Data links** section of the right menu.

Reviewer 2: ```suggestion
1. In the right-side menu, under **Data links**, select **Add link**.
```

1. Add a **Title** for the data link.

> 1. Find the UUID of the dashboard by looking in your browser’s address bar when the full dashboard is being rendered. Because this is a link to a dashboard in the same Grafana instance, only the path of the dashboard is required. <br /> {{< figure src="/static/img/docs/panels/traces/screensnot-traces-uuid-url.png" caption="Unique identifier for the dashboard" >}}

Reviewer 2: ```suggestion
1. Find the UUID of the dashboard by looking in your browser’s address bar when the full dashboard is being rendered. Because this is a link to a dashboard in the same Grafana stack, only the path of the dashboard is required. <br /> {{< figure src="/static/img/docs/panels/traces/screensnot-traces-uuid-url.png" caption="Unique identifier for the dashboard" >}}
```

> 1. In the `URL` edit field, we are going to make self-reference to the dashboard that contains both of the panels. This self-reference uses the value of the selected trace in the table to fill in the dashboard variable. We’ll use the path for the dashboard from the previous step and then fill in the value of `traceId` using the selected results from the TraceQL table. The trace ID is exposed using the `traceID` data field in the returned results, so we use that as the value for our dashboard variable. <br/> {{< figure class="float-right"  src="/static/img/docs/panels/traces/screenshot-traces-edit-link.png" caption="Edit link and add the Trace link" >}}

Reviewer 2: Changing this all to second person
```suggestion
1. In the **URL** field, make a self-reference to the dashboard that contains both of the panels. This self-reference uses the value of the selected trace in the table to fill in the dashboard variable. Use the path for the dashboard from the previous step and then fill in the value of `traceId` using the selected results from the TraceQL table. The trace ID is exposed using the `traceID` data field in the returned results, so use that as the value for the dashboard variable. <br/> {{< figure class="float-right"  src="/static/img/docs/panels/traces/screenshot-traces-edit-link.png" caption="Edit link and add the Trace link" >}}
```

1. Select **Save** to save the data link.
1. Select **Apply** from the panel editor to apply the panel to the dashboard.

>    Extend the width of the new table panel so that it covers the width of your dashboard.

Reviewer 2: ```suggestion
```

>    It’s also worth saving your dashboard at this point!

Reviewer 2: ```suggestion
1. Save the dashboard.
```

> You should now see a list of matching traces in the Table visualization. While selecting the `TraceID` or `SpanID` fields will give you the option to either open an Explorer page to visualize the trace or following the data link, selecting any other field (such as `Start time`, `Name` or `Duration`) automatically follows the data link, filling in the `traceId` dashboard variable, and then shows the relevant trace in the trace panel.

Reviewer 2: Since these are now fields in the UI, giving them the same UI element formatting.
```suggestion
You should now see a list of matching traces in the Table visualization. While selecting the **TraceID** or **SpanID** fields will give you the option to either open the **Explore** page to visualize the trace or following the data link, selecting any other field (such as **Start time**, **Name** or **Duration**) automatically follows the data link, filling in the `traceId` dashboard variable, and then shows the relevant trace in the trace panel.
```

Author: Thank you.

# Traces panel

Traces are data that lets you follow a request as it traverses the services in your infrastructure.
The Traces panel visualizes this data into a diagram that allows you to easily interpret this data.

For more information about traces and how to use them, refer to the following documentation:

- [What are traces](/docs/grafana-cloud/traces)
- [Tracing in Explore]({{< relref "../../../explore/trace-integration/" >}})
- [Tempo data source]({{< relref "../../../datasources/tempo/" >}})
- [Getting started with Tempo](/docs/tempo/latest/getting-started)

{{< figure src="/static/img/docs/explore/trace-view-9-4.png" class="docs-image--no-shadow" max-width= "900px" caption="Screenshot of the trace view" >}}

## Add a panel with tracing visualizations

Once you have tracing data available in your Grafana instance, you can add tracing panels to your Grafana dashboards.

> Using a dashboard variable, `traceID`, lets you create a query to show specific traces for a given Trace ID.

Reviewer: ```suggestion
Using a dashboard variable, `traceID`, lets you create a query to show specific traces for a given trace ID.
```

### Before you begin

To use this procedure, you need:

- A Grafana instance
- A Tempo data source connected to your Grafana instance (see [documentation](/docs/grafana/datasources/tempo/))

### Add the Traces panel query

To view and analyze traces data in a dashboard, you need to add the tracing panel to your dashboard and define a query using the panel editor.
The query determines the data that is displayed in the panel.
For more information on the panel editor, refer to the [Panel editor documentation]({{< relref "../../panel-editor-overview" >}}).

> This procedure uses dashboard variables and templates to allow us to enter trace IDs which can then be visualized. We’re going to use a variable called `traceId` and add it as a template query.

Reviewer 2: ```suggestion
This procedure uses dashboard variables and templates to allow you to enter trace IDs which can then be visualized. You'll use a variable called `traceId` and add it as a template query.
```

> 1. From your Grafana instance, create a new dashboard or go to an existing dashboard where you would like to add tracing panels.

Reviewer 2: ```suggestion
1. From your Grafana stack, create a new dashboard or go to an existing dashboard where you'd like to add tracing panels.
```

1. Select **Add visualization** from a new dashboard or select **Add Panel** on an existing dashboard.
1. Search for and select the appropriate tracing data source.

> 1. In the panel search pane on the right, select the **Visualizations** tab, search for, and select **Traces**.

Reviewer 2: I don't think we call this the search pane and the Viz picker is above the search bar, so I'm thinking just go with a more directional reference. 
```suggestion
1. In the top-right of the panel editor, select the **Visualizations** tab, search for, and select **Traces**.
```

1. In the panel search pane on the right, select the **Visualizations** tab, search for and select for Traces.
1. On the **Panel editor**, enter a **Title** for your trace panel. For more information on the panel editor, see the [Panel editor documentation]({{< relref "../../panels-visualizations/panel-editor-overview" >}}).
1. Select the **TraceQL** query type tab from the query editor.
1. Enter `${traceId}` in the TraceQL query field to create a dashboard variable. This variable is used as the template query.
   {{< figure class="float-right"  src="/static/img/docs/panels/traces/screenshot-traces-template-query.png" caption="Add a template query" >}}

> 

Author: The numbering continues. Unfortunately, new line with the figure shortcode causes the image to have a number. I'll fix that.

> 

Author: Numbers and images fixed.

1. Select **Apply** in the Panel editor to add the panel to the dashboard.
1. Expand the size of the trace panel so that it fills the width of the dashboard. We’ll now use the dashboard settings to create the variable used in the panel’s template query.
1. Go to the dashboard **Settings** and add a new variable called `traceId`, of variable type **Custom**, giving it a label if required. Select **Apply** to add the variable to the dashboard.
  {{< figure class="float-right"  src="/static/img/docs/panels/traces/screenshot-traces-custom-variable.png" caption="Add a Custom variable" >}}

> 1. You can verify that the panel works by using a valid Trace ID for the data source used for the trace panel and editing the ID in the dashboard variable.

Reviewer: ```suggestion
1. You can verify that the panel works by using a valid trace ID for the data source used for the trace panel and editing the ID in the dashboard variable.
```

   {{< figure src="/static/img/docs/panels/traces/screenshot-traces-traceid-panel.png" caption="Results of query in Trace panel" >}}

## Add TraceQL with table visualizations

While you can add a trace panel visualization to a dashboard, having to manually add trace IDs as a dashboard variable is cumbersome.
It’s more useful to instead be able to use TraceQL queries to search for specific types of traces and then select appropriate traces from matching results.

1. In the same dashboard where you added the trace visualization, select **Add panel** to add a new visualization panel.
1. Select the same trace data source you used in the previous section.

> 1. In the panel search pane, select the **Visualizations** tab and search for **Table**.

Reviewer 2: ```suggestion
1. In the top-right of the panel editor, select the **Visualizations** tab, search for, and select **Table**.
```

1. In the query editor, select the **TraceQL** tab.
1. Under the **Panel options**, enter a **Title** for your trace panel.

> 1. Add an appropriate TraceQL query to search for traces that you would like to visualize in the dashboard. We’ll use a simple, static query as an example. You can write the TraceQL query as a template query to take advantage of other dashboard variables, if they exist. This lets you create dynamic queries based on these variables.

Author: ```suggestion
1. Add an appropriate TraceQL query to search for traces that you would like to visualize in the dashboard. This example uses a simple, static query. You can write the TraceQL query as a template query to take advantage of other dashboard variables, if they exist. This lets you create dynamic queries based on these variables.
```

   {{< figure src="/static/img/docs/panels/traces/screenshot-traces-dynamic-query.png" caption="Create a dynamic query" >}}

When results are returned from a query, the results are rendered in the panel’s table.

{{< figure src="/static/img/docs/panels/traces/screenshot-traces-returned-query.png" caption="Results of a returned query in the panel table" >}}

### Use a variable to add other links to traces

The results in the Traces panel include links to the **Explore** page that renders the trace. You can add other links to traces in the table that fill in the `traceId` dashboard variable when selected, so that the trace is visualized in the same dashboard.

> To do so, you need to create a set of data links in the panel. Take the following steps:

Author: ```suggestion
To create a set of data links in the panel, use the following steps:
```
